### DISCUSSION TITLE
Add built-in search to generated documentation

### DISCUSSION THREAD

**Discussion by michaeljones** [gleam-lang/gleam#1342] [Ideas & suggestions]
To allow for easily finding a function or type within the modules of the package.

Possible inspiration:
- Elm generated docs: https://package.elm-lang.org/packages/elm/core/latest/
- mdbook search: https://rust-lang.github.io/mdBook/
- Rust generated docs search: https://doc.rust-lang.org/std/

The particular motivation was wanting to find if there was a 'maybe' or 'optional' style decoder in the dynamic package and just wanting to type those words into a search bar instead of navigate.

**Comment by michaeljones** [comment]
Some approaches would be:
- Simple substring matching (type & function names or maybe descriptions too)
- Fuzzy matching (levenstein distance type thing, type & function names or maybe descriptions too)
- Full text search (whatever that means - mdbook seems to use http://elasticlunr.com/)

Also whether to filter down the modules in the side bar or have some kind of pop-up list of results or some kind of full results page. 

I think it is also interesting to explore alternative names for things. Maybe you'd label the Result type with an `Either` tag/alternative-name as some people might look for that. (I actually did that when started with Elm and got quite confused.) Helps capture the idea that there are sometimes more than one way to name something and it improves discoverability to cover all options. Would involve some kind of tagging syntax in the comments though. Maybe for another day.

**Comment by lpil** [comment]
Fab idea, thank you very much.

> I think it is also interesting to explore alternative names for things. Maybe you'd label the Result type with an Either tag/alternative-name as some people might look for that. (I actually did that when started with Elm and got quite confused.) Helps capture the idea that there are sometimes more than one way to name something and it improves discoverability to cover all options. Would involve some kind of tagging syntax in the comments though. Maybe for another day.

This would be great too. We would need to think about how to add this information to the code.
